src/main/java/com/example/gradecalculator/service/SubjectService.java

```diff
@@ -0,0 +1,153 @@
+    }
+
+    public void removeSubjectForYear(int year, String subject) {
+        Set<String> selectedSubjects = selectedSubjectsByYear.get(year);
+        if (selectedSubjects != null) {
+            selectedSubjects.remove(subject);
+            if (selectedSubjects.isEmpty()) {
+                selectedSubjectsByYear.remove(year);
+            }
+        }
+    }
+
+    public List<String> saveSubjects(String[] selectedValues, long userId) {
+
+        var errors = new ArrayList<String>();
+        for (String entry : selectedValues) {
+            var splittedString = entry.split("-");
+            long selectedSubjectId = Long.parseLong(splittedString[0]);
+            long selectedYearId = Long.parseLong(splittedString[1]);
+            String error = saveSubjectYearSelectedUser(userId ,selectedSubjectId, selectedYearId);
+            if(error != null){
+                errors.add(error);
+            }
+        }
+
+        return errors;
+    }
+
+    public String saveSubjectYearSelectedUser(long userId, long subjectId, long yearId) {
+
+        Subject selectedSubject = subjectRepository.findById(subjectId).orElseThrow(() -> new IllegalArgumentException("Subject not found"));
+        SchoolYear selectedYear = schoolYearRepository.findById(yearId).orElseThrow(() -> new IllegalArgumentException("Year not found"));
+        User selectedUser = userRepository.findById(userId).orElseThrow(() -> new IllegalArgumentException("User not found"));
+        this.selectSubjectForYear(selectedYear.getStartDate().getYear(), selectedSubject.getName());
+        var subjects = userSubjectRepository.findByUserId(userId);
+        boolean exists = subjects.stream().anyMatch(item -> item.getSubject().getId().equals(subjectId) && item.getSchoolYear().getId().equals(yearId));
+        if (!exists) {
+            UserSubject userSubject = new UserSubject();
+            userSubject.setUser(selectedUser);
+            userSubject.setSubject(selectedSubject);
+            userSubject.setSchoolYear(selectedYear);
+            userSubjectRepository.save(userSubject);
+            return null;
+        }
+        return ("Subject already selected for the given year: " + selectedSubject.getName() + " - " + selectedYear.getName());
+    }
+
+    public TreeMap<String, Set<UserSubjectTO>> selectedSubject(long userId) {
+
+        var userSubjects = userSubjectRepository.findByUserId(userId);
+        TreeMap<String, Set<UserSubjectTO>> subjectsByYear = new TreeMap<>();
+        for (UserSubject userSubject : userSubjects) {
+            String year = userSubject.getSchoolYear().getName();
+            var userSubjectTO = new UserSubjectTO();
+            userSubjectTO.setID(userSubject.getId());
+            userSubjectTO.setName(userSubject.getSubject().getName());
+            subjectsByYear.computeIfAbsent(year, k -> new HashSet<>()).add(userSubjectTO);
+        }
+
+        return subjectsByYear;
+    }
+
+
+    @Transactional
```

Comment:
why did you choose to use this attribute?

Reply:
it helps successfully delete the subject and interact with DB

Reply:
first of all, I started using it cuz I could delete subject on view but not in DB. And it helped particularly in this case. And delete transaction gets rolled back if any exception occurs or if it fails using this annotation

Reply:
then it would be OK for me to use it here